sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs

```diff
@@ -163,35 +176,60 @@ data Error
   | EWrongInterfaceRequirement !(Qualified TypeConName) !(Qualified TypeConName)
-  | EUpgradeError !UpgradeError
   | EUnknownExperimental !T.Text !Type
+  | EUpgradeMissingModule !ModuleName
+  | EUpgradeMissingTemplate !TypeConName
+  | EUpgradeMissingChoice !ChoiceName
+  | EUpgradeMissingDataCon !TypeConName
+  | EUpgradeMismatchDataConsVariety !TypeConName
+  | EUpgradeRecordFieldsMissing !UpgradedRecordOrigin
+  | EUpgradeRecordFieldsExistingChanged !UpgradedRecordOrigin
+  | EUpgradeRecordFieldsNewNonOptional !UpgradedRecordOrigin
+  | EUpgradeRecordFieldsOrderChanged !UpgradedRecordOrigin
+  | EUpgradeVariantAddedVariant !UpgradedRecordOrigin
+  | EUpgradeVariantRemovedVariant !UpgradedRecordOrigin
+  | EUpgradeVariantChangedVariantType !UpgradedRecordOrigin
+  | EUpgradeVariantAddedVariantField !UpgradedRecordOrigin
+  | EUpgradeVariantVariantsOrderChanged !UpgradedRecordOrigin
+  | EUpgradeEnumAddedVariant !UpgradedRecordOrigin
+  | EUpgradeEnumRemovedVariant !UpgradedRecordOrigin
+  | EUpgradeEnumVariantsOrderChanged !UpgradedRecordOrigin
+  | EUpgradeRecordChangedOrigin !TypeConName !UpgradedRecordOrigin !UpgradedRecordOrigin
+  | EUpgradeTemplateChangedKeyType !TypeConName
+  | EUpgradeChoiceChangedReturnType !ChoiceName
+  | EUpgradeTemplateRemovedKey !TypeConName !TemplateKey
+  | EUpgradeTemplateAddedKey !TypeConName !TemplateKey
+  | EUpgradeTriedToUpgradeIface !TypeConName
+  | EUpgradeMissingImplementation !TypeConName !TypeConName
+  deriving (Show)
 
-data UpgradeError
-  = MissingModule !ModuleName
-  | MissingTemplate !TypeConName
-  | MissingChoice !ChoiceName
-  | MissingDataCon !TypeConName
-  | MismatchDataConsVariety !TypeConName
-  | RecordFieldsMissing !UpgradedRecordOrigin
-  | RecordFieldsExistingChanged !UpgradedRecordOrigin
-  | RecordFieldsNewNonOptional !UpgradedRecordOrigin
-  | RecordFieldsOrderChanged !UpgradedRecordOrigin
-  | VariantAddedVariant !UpgradedRecordOrigin
-  | VariantRemovedVariant !UpgradedRecordOrigin
-  | VariantChangedVariantType !UpgradedRecordOrigin
-  | VariantAddedVariantField !UpgradedRecordOrigin
-  | VariantVariantsOrderChanged !UpgradedRecordOrigin
-  | EnumAddedVariant !UpgradedRecordOrigin
-  | EnumRemovedVariant !UpgradedRecordOrigin
-  | EnumVariantsOrderChanged !UpgradedRecordOrigin
-  | RecordChangedOrigin !TypeConName !UpgradedRecordOrigin !UpgradedRecordOrigin
-  | TemplateChangedKeyType !TypeConName
-  | ChoiceChangedReturnType !ChoiceName
-  | TemplateRemovedKey !TypeConName !TemplateKey
-  | TemplateAddedKey !TypeConName !TemplateKey
-  | TriedToUpgradeIface !TypeConName
-  | MissingImplementation !TypeConName !TypeConName
-  deriving (Eq, Ord, Show)
+data WarnableError
+  = WEUpgradeShouldDefineIfacesAndTemplatesSeparately
+  | WEUpgradeShouldDefineIfaceWithoutImplementation !TypeConName ![TypeConName]
+  | WEUpgradeShouldDefineTplInSeparatePackage !TypeConName !TypeConName
+  | WEWarningToError !Warning
```

Comment:
Could you lift this case up to the `Error` type, to avoid the mutual recursion?

Reply:
Good idea! Fixed